### PR TITLE
feature: Add support for the dotcover format

### LINKED ISSUE #49: Add support to DotCover coverage format
This tool should be able to replace the tool located at https://github.com/codacy/csharp-codacy-coverage

### PR DESCRIPTION
**Pull request by franciscodua**
Fixes #49

### CODE REVIEW

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    } yield {
+      val filename = TextUtils.sanitiseFilename(fileIndices(fileIndex)).stripPrefix(projectRootStr).stripPrefix("/")
+
+      val lineCoverage = getLineCoverage(statements)
+      val totalLines = lineCoverage.keys.size
+      val coveredLines = lineCoverage.values.count(_ > 0)
+      val total = if (totalLines == 0) 0 else math.round((coveredLines.toFloat / totalLines) * 100)
+
+      CoverageFileReport(filename, total, lineCoverage)
+    }
```
Comment (lolgab) [inline src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: You can put all the `val`s in the for comprehension:
```suggestion
    val fileReports = for {
      (fileIndex, statements) <- statementsPerFile
      filename = TextUtils.sanitiseFilename(fileIndices(fileIndex)).stripPrefix(projectRootStr).stripPrefix("/")
      lineCoverage = getLineCoverage(statements)
      totalLines = lineCoverage.keys.size
      coveredLines = lineCoverage.values.count(_ > 0)
      total = if (totalLines == 0) 0 else math.round((coveredLines.toFloat / totalLines) * 100)
    } yield CoverageFileReport(filename, total, lineCoverage)    
```

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      case Success(_) =>
+        Left(s"Invalid report. Could not find tag <$RootTag $CoverageAttribute=...>.")
+
+      case Failure(ex) =>
+        Left(s"Unparseable report. ${ex.getMessage}")
+    }
+  }
+
+  private def hasCorrectSchema(elem: Elem) =
+    (elem \\ RootTag \ s"@$CoverageAttribute").nonEmpty
```
Comment (lolgab) [inline src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: You can use attribute method:
```suggestion
    (elem \\ RootTag \@ CoverageAttribute).nonEmpty
```

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  private def getLineCoverage(statementNodes: NodeSeq) = {
+    val lines = for {
+      node <- statementNodes
+      // a statement can extend over several lines
+      line <- (node \@ "Line").toInt to (node \@ "EndLine").toInt
+    } yield {
+      val coveredValue = if ((node \@ CoveredAttribute).toBoolean) 1 else 0
+      (line, coveredValue)
+    }
```
Comment (lolgab) [inline src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Same here:
```suggestion
      coveredValue = if ((node \@ CoveredAttribute).toBoolean) 1 else 0
    } yield (line, coveredValue)
```

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala (hunk 2 above)
Comment (franciscodua) [reply src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: From what I saw, doing `x \@ "..."` is the same as doing `(x \ "@...").text`
In this scenario I don't really want the text part. Just the NodeSeq

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala (hunk 1 above)
Comment (franciscodua) [reply src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Done.
Great tip

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala (hunk 3 above)
Comment (franciscodua) [reply src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Done

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    val fileReports = for {
+      (fileIndex, statements) <- statementsPerFile
+      filename = TextUtils.sanitiseFilename(fileIndices(fileIndex)).stripPrefix(projectRootStr).stripPrefix("/")
+      lineCoverage = getLineCoverage(statements)
+      totalLines = lineCoverage.keys.size
+      coveredLines = lineCoverage.values.count(_ > 0)
+      total = if (totalLines == 0) 0 else math.round((coveredLines.toFloat / totalLines) * 100)
+    } yield {
+      CoverageFileReport(filename, total, lineCoverage)
+    }
```
Comment (josemiguelmelo) [inline src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Remove brackets

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+  private def getLineCoverage(statementNodes: NodeSeq) = {
+    val lines = for {
+      node <- statementNodes
+      // a statement can extend over several lines
+      line <- (node \@ "Line").toInt to (node \@ "EndLine").toInt
+      coveredValue = if ((node \@ CoveredAttribute).toBoolean) 1 else 0
+    } yield {
+      (line, coveredValue)
+    }
```
Comment (josemiguelmelo) [inline src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Same here

Comment (franciscodua) [reply src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Done

File: src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala (hunk 4 above)
Comment (franciscodua) [reply src/main/scala/com/codacy/parsers/implementation/DotcoverParser.scala]: Done

**Pull request merged**